[WebGPU] copyToDestinationFormat copies too many bytes for Rg32float

mwyrzykowski opened the pull request:
#### ee1d490ee70c84af0993d2420bb77088c760a0b2
<pre>
[WebGPU] copyToDestinationFormat copies too many bytes for Rg32float
<a href="https://bugs.webkit.org/show_bug.cgi?id=270203">https://bugs.webkit.org/show_bug.cgi?id=270203</a>
&lt;radar://122405464&gt;

Reviewed by Dan Glastonbury.

This code will go away when <a href="https://bugs.webkit.org/show_bug.cgi?id=263692">https://bugs.webkit.org/show_bug.cgi?id=263692</a> is
implemented but for now fix the out of bounds access which was caught
by ASAN.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::copyToDestinationFormat):

Canonical link: <a href="https://commits.webkit.org/275431@main">https://commits.webkit.org/275431@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/065cf3000a525e64abba36649f02de4a8a8c0541

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41794 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20808 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44176 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44372 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37887 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44101 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23985 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18139 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34541 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42368 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17760 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35994 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15211 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15448 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37018 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45766 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38003 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37340 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41090 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16609 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13640 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39536 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18228 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9370 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18286 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17872 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
```diff
@@ -578,7 +578,7 @@ static void* copyToDestinationFormat(const uint8_t* rgbaBytes, GPUTextureFormat
             data[i0 + 1] = rgbaBytes[i + 1];
         }
 
-        sizeInBytes = sizeInBytes * sizeof(float);
```
see malloc call on line 575, I thought about making a wrapper function but sizeInBytes is used in all the for loops

djg commented on Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
```diff
@@ -578,7 +578,7 @@ static void* copyToDestinationFormat(const uint8_t* rgbaBytes, GPUTextureFormat
         }
 
-        sizeInBytes = sizeInBytes * sizeof(float);
+        sizeInBytes = (sizeInBytes / 2) * sizeof(float);
```
Maybe this value should be lifted to Line 575 and named `texSizeInBytes` and used in the malloc and change this to `sizeInBytes = texSizeInBytes`?

djg replied:
Or `dataSizeInBytes` (naming is hard)

mwyrzykowski replied:
I considered that and think it's a good idea. However, this same code occurs about 8 times in this function, I think the risk of accidentally introducing a regression by cleaning it up outweighs the benefit of cleaning it up, as it will be removed entirely once  https://bugs.webkit.org/show_bug.cgi?id=263692  is completed.

djg reviewed (approved):
:shipit:

mwyrzykowski committed:
[WebGPU] copyToDestinationFormat copies too many bytes for Rg32float
https://bugs.webkit.org/show_bug.cgi?id=270203
<radar://122405464>

Reviewed by Dan Glastonbury.

This code will go away when https://bugs.webkit.org/show_bug.cgi?id=263692 is
implemented but for now fix the out of bounds access which was caught
by ASAN.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::copyToDestinationFormat):

Canonical link: https://commits.webkit.org/275431@main
